fmlearlydisplay/src/main/java/net/minecraftforge/fml/earlydisplay/DisplayWindow.java

```diff
@@ -0,0 +1,531 @@
+        try (MemoryStack memorystack = MemoryStack.stackPush()) {
+            PointerBuffer pointerbuffer = memorystack.mallocPointer(1);
+            int error = org.lwjgl.glfw.GLFW.glfwGetError(pointerbuffer);
+            if (error != GLFW_NO_ERROR) {
+                long pDescription = pointerbuffer.get();
+                String description = pDescription == 0L ? "" : MemoryUtil.memUTF8(pDescription);
+                handler.accept(error, description);
+            }
+        }
+    }
+
+    /**
+     * Hand-off the window to the vanilla game.
+     * Called on the main thread instead of the game's initialization.
+     *
+     * @return the Window we own.
+     */
+    public long setupMinecraftWindow(final IntSupplier width, final IntSupplier height, final Supplier<String> title, final LongSupplier monitorSupplier) {
+        // we have to spin wait for the window ticker
+        while (!this.windowTick.isDone()) {
+            this.windowTick.cancel(false);
+        }
+        try {
+            renderLock.tryAcquire(100, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Thread.interrupted();
+        }
+        // schedule a 50 ms ticker to try and smooth out the rendering
+        renderScheduler.scheduleAtFixedRate(()->tickSemaphore.set(true), 50, 50, TimeUnit.MILLISECONDS);
+
+        glfwMakeContextCurrent(window);
+        // Set the title to what the game wants
+        glfwSetWindowTitle(window, title.get());
+        glfwSwapInterval(0);
+        // Clean up our hooks
+        glfwSetFramebufferSizeCallback(window, null).free();
+        glfwSetWindowPosCallback(window, null).free();
+        glfwSetWindowSizeCallback(window, null).free();
+        return window;
+    }
+
+    @Override
+    public boolean positionWindow(final Optional<Object> monitor, final IntConsumer widthSetter, final IntConsumer heightSetter, final IntConsumer xSetter, final IntConsumer ySetter) {
+        widthSetter.accept(this.winWidth);
+        heightSetter.accept(this.winHeight);
+        xSetter.accept(this.winX);
+        ySetter.accept(this.winY);
+        return true;
+    }
+
+    @Override
+    public void updateFramebufferSize(final IntConsumer width, final IntConsumer height) {
+        width.accept(this.fbWidth);
+        height.accept(this.fbHeight);
+    }
+
+    private Method loadingOverlay;
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> Supplier<T> loadingOverlay(final Supplier<?> mc, final Supplier<?> ri, final Consumer<Optional<Throwable>> ex, final boolean fade) {
+        try {
+            return (Supplier<T>)loadingOverlay.invoke(null, mc, ri, ex, this);
+        } catch (Throwable e) {
+            throw new IllegalStateException("How did you get here?", e);
+        }
+    }
+
+    @Override
+    public void updateModuleReads(final ModuleLayer layer) {
+        var fm = layer.findModule("forge").orElseThrow();
```

Comment:
This hardcoded module lookup expects `forge` to always be present, which is not the case in `fmlonly` environments, leading to a crash. We should find a way to detect which environment we're in and determine the correct module name.

Reply:
This will only load in specific environments where forge should always be available. Fmlonly won't support this because the changes are too intrusive to implement using that system.

Reply:
Ok. The NoVizCallback is a problem. It needs to live in Minecraft world but is dynamically loaded. I'll see if I can find a better place.

Reply:
Fixed this. NoVizCallback is unnecessary in fmlonly, because LoadingOverlay is unpatched, so we just skip it if the "forge" module isn't present.